### PR TITLE
♻️ replace deprecated functions for communication

### PR DESCRIPTION
**Pull request by mayank31398**
This PR is for using the new communication functions
`all_gather_into_tensor` instead of `_all_gather_base`
`reduce_scatter_tensor` instead of `_reduce_scatter_base`

If these functions are not found (for older torch versions), we default to the deprecated functions.

### CODE REVIEW

**Comment by mayank31398** [comment]
@microsoft-github-policy-service agree

**Comment by mayank31398** [comment]
> @mayank31398 please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by mayank31398** [comment]
@jeffra @RezaYazdaniAminabadi , I am unsure why the test is failing here.
The is CUDA OOM in `nv-torch18-v100 / unit-tests`

File: deepspeed/comm/comm.py
Comment (Quentin-Anthony) [inline deepspeed/comm/comm.py:337]: This is not backwards-compatible with older versions of torch that should still be using `all_gather_base`. I think a better structure would be:

```
if cdb.has_all_gather_into_tensor():
    return all_gather_into_tensor(...)
elif cdb.has_allgather_base():
    return all_gather_base(...)
else:
    # print warning on all_gather perf
    return all_gather(...)
```

it'd be cleaner to use your `get_all_gather_function()` here and only print the warning if `get_all_gather_function is None`

File: deepspeed/comm/comm.py
Comment (Quentin-Anthony) [inline deepspeed/comm/comm.py:266]: Same comment on backwards compatibility as for all_gather

File: benchmarks/communication/all_gather.py
Comment (Quentin-Anthony) [inline benchmarks/communication/all_gather.py:20]: The reason for this runtime-check is to enable users to benchmark communication ops using pure `torch.distributed` even without deepspeed installed (e.g. users of NeMo, Megatron-LM, fairseq, ColossalAI, etc). 

I'm of the opinion that we should preserve this behavior for the comms benchmarks (i.e. no `deepspeed.*` top-level imports and wrap all deepspeed- or torch.distributed-specific code in an `if args.dist == ''` check), but I think @tjruwase and/or @jeffra have the final say. Can one of you provide comment?

File: deepspeed/comm/comm.py
Comment (mayank31398) [reply deepspeed/comm/comm.py:266]: same comment as above

File: deepspeed/comm/comm.py
Comment (mayank31398) [reply deepspeed/comm/comm.py:337]: I think it should work with older versions too right?
Screenshot 2023-03-15 at 3 27 15 PM
this populates the function and for older versions it should work.

Am i missing something?

Comment (mayank31398) [reply deepspeed/comm/comm.py:337]: I am populating the all_gather_function here:
https://github.com/mayank31398/DeepSpeed/blob/3133caff25307d63dc248b02416420eb9288189d/deepspeed/comm/torch.py#L28

Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py:337]: Ah I see. The naming of `has_all_gather_into_tensor` is a bit confusing since this checks for both `all_gather_into_tensor` and `_all_gather_base`. 

I think https://github.com/microsoft/DeepSpeed/pull/3020 introduces a more clear naming convention with `allgather_base_fn`

Comment (mayank31398) [reply deepspeed/comm/comm.py:337]: I agree :)

**Comment by stas00** [comment]
if this one is a go please force a CI restart - HF side has been fixed.

File: benchmarks/communication/all_gather.py
Comment (tjruwase) [reply benchmarks/communication/all_gather.py:20]: Looks good to me.

File: deepspeed/comm/comm.py
Comment (tjruwase) [reply deepspeed/comm/comm.py:330]: @mayank31398, do you plan to adapt the naming convention to #3020? I am fine either way.

**Review by tjruwase — APPROVED** [review]

**Comment by mayank31398** [comment]
Not sure why the CI is crashing @tjruwase .
Doesn't seem to be a bug on my end.
I will restart the CI build again once.

**Comment by mayank31398** [comment]
CI still seems to be broken :(

**Comment by jeffra** [comment]
> CI still seems to be broken :(

Sorry for delays, we're actively working on trying to unblock several CI related issues. Once this PR is merged we should be good to go: https://github.com/microsoft/DeepSpeed/pull/3047

File: deepspeed/comm/comm.py
Comment (mayank31398) [reply deepspeed/comm/comm.py:330]: not really, the new convention in torch is `gather_into_tensor`, so it would be better if we adopt this sooner?

**Comment by mayank31398** [comment]
@jeffra @tjruwase the tests are still crashing, seems unrelated to this PR.
Is the transformers version correct?

**Comment by mayank31398** [comment]
Successful checks @jeffra @tjruwase

**Comment by mayank31398** [comment]
its broken again :|

**Comment by stas00** [comment]
yeah, looks like something broke in transformers again - looking

**Comment by stas00** [comment]
I'm able to reproduce once I install `safetensors` - it works without it - investigating.

**Comment by stas00** [comment]
@jeffra, the transformers tests are unrelated to this feature. It should be safe to merge. @mayank31398 has been waiting for eons for this to be merged. Thank you!

**Comment by mayank31398** [comment]
Thanks Stas :)

**Comment by stas00** [comment]
and the transformers job should be fine now if someone can trigger the rebuild.

**Comment by tjruwase** [comment]
> Can we prioritize this one? @tjruwase :)

Yes, will merge once this CI run completes.

**Comment by tjruwase** [comment]
@mayank31398, I think the formatting issues can be fixe by upgrading pre-commit and clang-format

File: benchmarks/communication/all_gather.py
Comment (Quentin-Anthony) [inline benchmarks/communication/all_gather.py:3]: These imports were moved down to https://github.com/microsoft/DeepSpeed/pull/2995/files#diff-57ff0beda6888da8839b4b96071460313d9ad4d5b6ce148337374409602b9403R11. 

`from deepspeed.comm import TorchBackend` should be moved to L14, and the rest in this block can be deleted.

Comment (mayank31398) [reply benchmarks/communication/all_gather.py:3]: > @mayank31398, I think the formatting issues can be fixe by upgrading pre-commit and clang-format

i am not seeing any issues with the formatting in the CI.
are you suggesting to update the pre-commit nonetheless @tjruwase ?

Comment (Quentin-Anthony) [reply benchmarks/communication/all_gather.py:3]: It's still off. The beginning of this file should look like:

```
'''Copyright The Microsoft DeepSpeed Team'''

import torch
import sys, os, time

COMMS_BENCH_DIR = os.path.join(os.path.dirname(__file__), "../")
sys.path.append(COMMS_BENCH_DIR)

from communication.utils import *
from communication.constants import *
from deepspeed.accelerator import get_accelerator
from deepspeed.comm import TorchBackend
```

Comment (mayank31398) [reply benchmarks/communication/all_gather.py]: done

Comment (Quentin-Anthony) [reply benchmarks/communication/all_gather.py]: Great! I'm happy with this. Thanks for sticking through it @mayank31398 :)

Comment (mayank31398) [reply benchmarks/communication/all_gather.py]: Awesome, lets get this in. Don't wanna rebase against master again :)

**Pull request merged**
